src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts

```diff
@@ -132,4 +136,59 @@ export class TableStructureComponent implements OnChanges {
     return unique;
   }
 
+  classifyComplexesSimilaritiesV2(complexesList: Element[]) {
+    const comparedComplexes: [Element, Element, number][] = [];
+    const finalList: Element[] = [];
+
+    for (const complex of complexesList) {
+      for (const comparedComplex of complexesList) {
+        // for unique comparison
+        if (complex.complexAC >= comparedComplex.complexAC) {
+          comparedComplexes.push([complex, comparedComplex, this.calculateSimilarity(complex, comparedComplex)]);
+        }
+      }
+    }
+    comparedComplexes.sort((a, b) => b[2] - a[2]); // sorting by similarityScore
+
+    // const complexesOrderedSet = new Set<Element>(); // used to have unique complexes
+    // for (const [complex1, complex2] of comparedComplexes) {
+    //   if (!complexesOrderedSet.has(complex1)) {
+    //     complexesOrderedSet.add(complex1);
+    //   }
+    //   if (!complexesOrderedSet.has(complex2)) {
+    //     complexesOrderedSet.add(complex2);
+    //   }
+    // }
+
+    const complexesOrderedSet = this.uniqueComplexesListOrderedBySimilarity(comparedComplexes);
+    // to be used in the table as a 1D array
+    for (const complex of complexesOrderedSet) {
+      finalList.push(complex);
+    }
+    return finalList;
+  }
+
+  uniqueComplexesListOrderedBySimilarity(complexesListSimilarities: [Element, Element, number][]) {
```

Comment:
We have 2 for loops here where we iterate through the complexesListSimilarities list. It makes sense but it can be more slightly more efficient.

The second time we iterate we can start from where we are, as anything before that point has already been checked. To do that, we can use indexes in the for loop.

```
    for (let i = 0; i < complexesListSimilarities.length; i++) {
      const complex1 = complexesListSimilarities[i][0];
      const complex2 = complexesListSimilarities[i][1];
      const similarityScore = complexesListSimilarities[i][2];
      if (similarityScore !== 0) {
        if (!complexesOrderedSet.has(complex1)) {
          complexesOrderedSet.add(complex1);
        }
        if (!complexesOrderedSet.has(complex2)) {
          complexesOrderedSet.add(complex2);
        }
        for (let j = i + 1; j < complexesListSimilarities.length; j++) {
          const complex3 = complexesListSimilarities[j][0];
          const complex4 = complexesListSimilarities[j][1];
          const similarityScore2nd = complexesListSimilarities[j][2];
          if (complex1 === complex3 && similarityScore2nd !== 0) {
            if (!complexesOrderedSet.has(complex4)) {
              complexesOrderedSet.add(complex4);
            }
          }
        }
      }
    }
```

I think that would make this from O(n^2) to O(nLogn), and it should work the same way.

Reply:
Basically, if we have 4 complexes (CPX1, CPX2, CPX3 and CPX4), `complexesListSimilarities` could look like
```
complexesListSimilarities = [
  [CPX1, CPX1, 5],
  [CPX2, CPX2, 5],
  [CPX1, CPX2, 4],
  [CPX3, CPX3, 3],
  [CPX1, CPX4, 2],
  [CPX4, CPX4, 2],
  [CPX2, CPX3, 1],
  [CPX2, CPX4, 1],
  [CPX3, CPX4, 0],
  [CPX1, CPX3, 0]
]
```

First, on the first loop, with i=0, we check `[CPX1, CPX1, 5]` and we add CPX1 to the set.
Then, we do the second loop from j=1 (i+1), to j=length-1, and we add CPX2 and CPX4 to the set.

Then, i=1, CPX2 is already added to the set. The second loop now starts from j=2.
CPX3 is added to the set.

The rest of iterations on the loops will not add anything, so we end up with [CPX1, CPX2, CPX4, CPX3].